crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -952,6 +977,82 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
+    /// Check the overloaded functions in this scope.
+    ///
+    /// This only checks the overloaded functions that are:
+    /// 1. Visible publicly at the end of this scope
+    /// 2. Or, defined and called in this scope
+    ///
+    /// For (1), this has the consequence of not checking an overloaded function that is being
+    /// shadowed by another function with the same name in this scope.
+    fn check_overloaded_functions(&mut self) {
+        let mut functions_to_check = self.called_functions.clone();
+
+        // Collect all the unique overloaded function symbols in this scope. This requires a set
+        // because an overloaded function uses the same symbol for each of the overloads and the
+        // implementation.
+        let overloaded_function_symbols: FxHashSet<_> = self
+            .types
+            .declarations
+            .iter()
+            .filter_map(|(definition, ty)| {
+                // Filter out function literals that result from anything other than a function
+                // definition e.g., imports.
+                if !matches!(definition.kind(self.db()), DefinitionKind::Function(_)) {
+                    return None;
+                }
+                let function = ty.inner_type().into_function_literal()?;
+                if function.has_known_decorator(self.db(), FunctionDecorators::OVERLOAD) {
+                    Some(definition.symbol(self.db()))
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        let use_def = self
+            .index
+            .use_def_map(self.scope().file_scope_id(self.db()));
+
+        for symbol in overloaded_function_symbols {
+            if let Symbol::Type(Type::FunctionLiteral(function), Boundness::Bound) =
+                symbol_from_bindings(self.db(), use_def.public_bindings(symbol))
+            {
+                // Extend the functions that we need to check with the publicly visible overloaded
+                // function. This is always going to be either the implementation or the last
+                // overload if the implementation doesn't exists.
+                functions_to_check.insert(function);
+            }
+        }
+
+        for function in functions_to_check {
+            let Some(overloaded) = function.to_overloaded(self.db()) else {
+                continue;
+            };
+
+            let function_node = function.node(self.db());
```

Comment:
This line will create extra Salsa dependencies. Since we are taking care above to only include functions defined in this file, that should be OK, but it will still be more efficient if we only do this when we actually need it -- that is, inside the actual error condition `if` body below. Ideally also inside the `if let Some(builder) { ... }` so we don't do it if diagnostics are not enabled here.

Reply:
I'm not sure if it's possible to move it inside the `if let Some(builder) { ... }` branch because the `report_lint` requires the `Ranged` value which is the function name node. I don't think I can use `focus_range` because that also accesses the `node` in the same way as this method.